### PR TITLE
ARROW-11492: [Rust] Replaced many of the `Array::new` by `safe` counterparts

### PR DESCRIPTION
**Pull request by jorgecarleitao**
## Rational

When `ArrayData::new` is used, we make no attempt to verify that its contents lead to sound code. In particular, for primitive arrays, the following must hold:

```
/// * `values.len()` is a multiple of `size_of::<T::Native>`
/// * `values.as_ptr()` is aligned with `T::Native`
/// * when `nulls` is `Some`, `nulls.len` is equal to `(ArrayData::len + 7) / 8`.
/// * `offset <= ArrayData::len`
```

* The first two conditions allow us to soundly transmute the `u8` buffer into `T::Native`.
* The third condition allow us to access bits on the null bitmap up to `ArrayData::len`.
* The last condition avoids out of bound accesses to the data when offsets are used.

## This PR

This PR introduces a method to create an `ArrayData` for primitives by verifying that the above assumptions hold.

This PR also migrates most of the `ArrayData::new` for primitive arrays to it.

While migrating code to this version, I found an out-of-bound access in the `nullif` kernel derived from passing a value buffer whose `len` is smaller than the necessary. This is not fixed here.

# Notes

Note that are two aspects of unsafety here:

1. it is `unsafe` to pass a `len` to `ArrayData::new` that goes beyond `Buffer::len / sizeo_of<T>`. This is somewhat equivalent to the `TrustedLen` trait that Rust offers to "prove" that the `len` of an iterator is trustworthy.

2. it is `unsafe` to pass a buffer to `ArrayData::new` that does not match the corresponding `DataType`. Doing so is unsound but it is not possible to prove, by calling `ArrayData::new`, that transmuting a `Buffer` to `T::Native` is sound.

So, in general, `ArrayData::new` is _very_ `unsafe`. This PR is an attempt to replace some of these calls by a `safe` counterpart, that verifies the necessary invariants for primitive types.

I am in generally in favor of either
* mark `ArrayData::new` as `unsafe`  (backward incompatible API)
* force `new` to verify all assumptions (backward incompatible performace) and offer an `unsafe unchecked` version

It is expensive to verify that `ArrayData` is safe because e.g. for strings, we need to verify that the value buffer contains valid utf8s.

Generally, the hindrance here is that `ArrayData` supports all types, and this leads to all kinds of unsafety as it requires an untyped, non-generic `Buffer`. If we had a `Buffer<T>`, this would all be safe because we would not have to verify the buffers at runtime, either when we downcast arrays, or when we create arrays. Currently, we (unsafely) assume that `ArrayData` is correctly built and thus can transmute its buffers to type `T`. All very risky stuff.

### CODE REVIEW

**Comment by jorgecarleitao** [comment]
Any thoughts here, @alamb @nevi-me @vertexclique @troels @Dandandan @jhorstmann ?

**Comment by jhorstmann** [comment]
I think the compiler should be able to prove most of those assertions statically, so maybe we don't need an unsafe variant and could leave the assertions in.

For example, when the buffer is created via `from_trusted_len_iter`, that method calculates the length by multiplying with `size_of::<T>()` and allocates and writes into a suitable aligned buffer. This knowledge should allow the compiler to remove the corresponding assertions in `ArrayData::new_primitive`.

For further simplification, we could maybe remove the `offset` parameter here, I see it is 0 in all usages of this PR. If we leave it in, we should add a test with a non-zero offset, as I'm not convinced the logic is correct.

File: rust/arrow/src/array/data.rs
Comment (jhorstmann) [inline rust/arrow/src/array/data.rs:527]: Not sure this should substract the offset. The arrays offset is later uses to index into the logical elements of both null and data buffers, so they should have the same length (rounded up to full bytes for the null buffer).

Rather the len of the array below should be `len - offset`.

But even better would be if we don't need the offset parameter at all and avoid all this confusion ;)

File: rust/arrow/src/array/data.rs
Comment (yordan-pavlov) [inline rust/arrow/src/array/data.rs:544]: @jorgecarleitao in this PR the `offset` parameter is always 0; why not remove it in order to further simplify the code?

File: rust/arrow/src/array/data.rs
Comment (jorgecarleitao) [reply rust/arrow/src/array/data.rs]: Good point. I removed the offset. 👍

File: rust/arrow/src/array/data.rs
Comment (jorgecarleitao) [reply rust/arrow/src/array/data.rs]: Good point. I removed the offset. 👍

**Comment by jorgecarleitao** [comment]
> I think the compiler should be able to prove most of those assertions statically, so maybe we don't need an unsafe variant and could leave the assertions in.
> 
> For example, when the buffer is created via `from_trusted_len_iter`, that method calculates the length by multiplying with `size_of::<T>()` and allocates and writes into a suitable aligned buffer. This knowledge should allow the compiler to remove the corresponding assertions in `ArrayData::new_primitive`.

That is a good hypothesis. I will do some benchmarking to test it, and if it holds, I will replace them.

**Comment by jorgecarleitao** [comment]
You were right, @jhorstmann , it has no significant impact and I thus removed the `unsafe` version and made all checks happen. Thanks a lot for that idea!

```
add 512                 time:   [1.0077 us 1.0126 us 1.0181 us]                     
                        change: [+5.0339% +6.3889% +7.6813%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  3 (3.00%) high severe

subtract 512            time:   [962.12 ns 968.88 ns 976.54 ns]                          
                        change: [+2.2544% +3.6466% +4.9816%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  6 (6.00%) high mild
  2 (2.00%) high severe

multiply 512            time:   [964.18 ns 967.85 ns 971.69 ns]                          
                        change: [-0.1040% +1.4543% +2.7701%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  2 (2.00%) high severe

divide 512              time:   [2.8418 us 2.8454 us 2.8492 us]                        
                        change: [-0.8891% -0.2418% +0.3257%] (p = 0.45 > 0.05)
                        No change in performance detected.
Found 12 outliers among 100 measurements (12.00%)
  2 (2.00%) low mild
  4 (4.00%) high mild
  6 (6.00%) high severe

limit 512, 512          time:   [400.88 ns 402.76 ns 404.82 ns]                           
                        change: [-2.9396% -1.4390% -0.1617%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

add_nulls_512           time:   [974.42 ns 976.84 ns 979.38 ns]                           
                        change: [-4.7016% -3.1536% -1.6052%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  4 (4.00%) high severe

divide_nulls_512        time:   [2.8035 us 2.8104 us 2.8175 us]                              
                        change: [-2.5459% -1.3916% -0.4130%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  4 (4.00%) high severe
```

**Comment by alamb** [comment]
This one looks like it has lots of eyes already, so I am going to defer to those reviewers unless you think it is also worth another look by me too -- I am struggling to keep up with so much (great) stuff going on

**Comment by alamb** [comment]
I didn't get a chance to review this today, but I plan to do so tomorrow

File: rust/arrow/src/compute/kernels/arithmetic.rs
Comment (alamb) [inline rust/arrow/src/compute/kernels/arithmetic.rs:228]: FWIW I think the new style of calling `ArrayData::new_primitive`  is much more understandable: Rather than 7 parameters which have some implicit assumptions about invariants between them, we have 2 parameters that make more sense.

File: rust/arrow/src/compute/kernels/boolean.rs
Comment (alamb) [inline rust/arrow/src/compute/kernels/boolean.rs:298]: I don't understand this comment -- is it just pointing out that a bug still exists in the code? It doesn't look like your change changes the behavior at this time. Maybe it is worth a JIRA ticket to track it

**Review by alamb — APPROVED** [review]
I like this change a lot @jorgecarleitao 👍 

I think it is good to go

**Comment by jorgecarleitao** [comment]
cc @nevi-me , related to #9492

File: rust/arrow/src/array/data.rs
Comment (nevi-me) [inline rust/arrow/src/array/data.rs:481]: Should this cater for offsets, or is your preference to omit it? If we took an offset, we'd add an extra assertion that checks that `offset < calculated_len`.

Thinking more about it, if we don't take an offset, we might force users to still use `ArrayData::new` for that facility, which might be undesirable.

File: rust/arrow/src/compute/kernels/arithmetic.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/arithmetic.rs:247]: I agree!

File: rust/arrow/src/compute/kernels/boolean.rs
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/boolean.rs:298]: There's 2 issues:

1. When slicing `ArrayData`, we do not pass the slice down to `Buffer`, even though it has a `slice()` method.

2. The way that we calculate `Buffer::len` is by taking its data (`Arc<Bytes>`) and subtracting `Buffer::offset`. The conclusion that I reach is that `Buffer::len()` is incorrect.

I came across this earlier today, and to address this, I'm trying out the following change:

```rust
pub struct Buffer {
    /// the internal byte buffer.
    data: Arc<Bytes>,

    /// The offset into the buffer.
    offset: usize,

    /// The logical length of the buffer. [ADDED]
    length: usize,
}
```

The only time we set `Buffer::offset` != 0 is on `Buffer::slice(offset: usize)`.
Now, if I have an array of 24 values, and slice it with offset 12 and length 8, we have a new array with len = 8, but its buffer's len is still 24. 
If we pass down both the offset and length to the buffer, we would make the whole thing consistent; and avoid the above check.

**Review by nevi-me — COMMENTED** [review]
Thanks @jorgecarleitao, this simplifies primitive cases a lot.

I'd like us to hash out the comment from @alamb on the primitive boolean a bit more. I think we have a bigger buffer offset & length soundness issue, which might take a few PRs to resolve.

How about we create a temporary branch, and merge buffer-related changes like this on it; then we can merge the completed changes into master after a few more PRs?

**Comment by nevi-me** [comment]
> For further simplification, we could maybe remove the `offset` parameter here, I see it is 0 in all usages of this PR. If we leave it in, we should add a test with a non-zero offset, as I'm not convinced the logic is correct.

@jhorstmann I went the opposite way to you on this. I was relying on my intuition; but perhaps I need to come with a test case to validate my thinking on whether we should leave the option of taking an offset open.

@jorgecarleitao we can keep the current behaviour for now.

File: rust/arrow/src/compute/kernels/boolean.rs
Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/boolean.rs:298]: @nevi-me , I reached that same conclusion independently! https://github.com/jorgecarleitao/arrow2/blob/proposal/src/buffer/immutable.rs#L37

I obviously agree with you. :)

Comment (nevi-me) [reply rust/arrow/src/compute/kernels/boolean.rs:298]: I'm going a step further, and adding a `bit_width` here.

```diff
pub struct Buffer {
    /// the internal byte buffer.
    data: Arc<Bytes>,

    /// The offset into the buffer.
    offset: usize,

+   /// The logical length of the buffer
+   length: usize,
+
+   bit_width: usize,
}
```

Your proposed implementation doesn't need it, because `NativeType` knows its byte width.

I think we should clarify that `offset` and `length` are independent of `Bytes` (or we should make them so).
For example, if I have Bytes::from(vec![0u8, 255u8]) as 16 `bool` values, the following relationship should exist:

```rust
offset + length <= 16;
data.len() == 2;
```

So, if we slice into the bytes, we need to know that the `bit_width` is 1, and an offset of 7 lies within the first byte of the data. Addressing this would also make our slices sound.

Comment (jorgecarleitao) [reply rust/arrow/src/compute/kernels/boolean.rs:298]: I was about to write that the slicing problem would not be fixed with this, but I see that you just proposed something to address it. I (again) fully agree with you that we need `bit_width` to fix slicing, particularly for the Boolean Arrays.

What I ended up doing in the other repo was to stop having bits handled in Buffers. I wrote a [`Bitmap`](https://github.com/jorgecarleitao/arrow2/blob/proposal/src/buffer/bitmap.rs#L13) that holds `Bytes`, and `MutableBitmap` that builds `Bytes` from bits / bool, and both are used exclusively for bit operations (everything is measured in bits). We can't to this here atm because all our arrays data resides in `ArrayData`, which means that the representation of the values of a `BooleanArray` must be a `Buffer`. This force us to use Buffers both from bits and byte-like types, which is what you are alluding to: the buffer must know what it contains.

fyi, the proposed implementation needs the `length` because there we are not even storing the length on the array, and instead use the same as in values:

```rust
#[derive(Debug, Clone)]
pub struct PrimitiveArray<T: NativeType> {
    data_type: DataType,
    values: Buffer<T>,
    validity: Option<Bitmap>,
    offset: usize,
}

impl<T: NativeType> Array for PrimitiveArray<T> {
    #[inline]
    fn len(&self) -> usize {
        self.values.len()
    }
}
```

This follows a similar pattern used by `Tokio-bytes`'s `Bytes` for memory sharing:

* `arrow::Bytes<T>` is an immutable and sharable memory region of type `T` that knows how to deallocate itself;
* `Buffer` is a slice `[start,end]` of one of those regions
* `Bitmap` is a bit slice `[start,end]` of `arrow::Bytes<u8>`

Note how `PrimitiveArray<T: NativeType>` cannot be initialized with an offset in my proposal: the only way to offset it is via `PrimitiveArray::slice`, and there we offset both validity and values consistently. This way, we never end up in an inconsistent state nor have to ask users to pass a consistent state.

Anyway, this just to re-state that I am in full agreement with you over this issue, and those are what led me to try out getting rid of `ArrayData` altogether: imo it is not allowing us to use structs that are convenient for each of the types. The problem is that changing this has a major impact in our code-base.

If we want to go through such a path, by now I know well what to do to migrate that to this repo, as I have been playing around on the other to understand the risks. I am convinced we can do it, but need a good orchestration because there are certain idioms in the current code base that we must stop using and remove. The simpler answer is that everything that uses an `ArrayData` must be replaced by calls to methods from the concrete implementations. Atm, that I noticed, these are

* equal/
* transform/
* slicing
* IPC
* FFI
* Parquet
* some of the JSON reader
* some cast implementations

One proposal:

1. ground work: re-write the crate in a concrete-type specific way, i.e. we need to basically remove `Array::data()` and `Array::data_ref()`, and replace them by concrete types' methods such as `values()`, `offsets()`, `fields()`, etc.
2. push the core of the proposal to a separate branch (say `proposal-rust2`), and PR it to a second branch, say `rust2`. This will trigger a formal discussion between us about it until we reach a consensus about it, at which point we merge it to `rust2`.
3. migrate each component to the new API.
4. Migrate DataFusion and Parquet.

This has the same / higher risks that the parquet writer branch had, though, but I can't really see an easier way of getting rid of `ArrayData`: our repo depends on it all the way from IPC and FFI to the filter kernel!

**Comment by alamb** [comment]
THis PR needs a rebase @jorgecarleitao 

@nevi-me  / @jhorstmann / @yordan-pavlov any other thoughts? Or shall we merge this one in?

**Comment by jhorstmann** [comment]
This looks like a clear improvement to me :+1: :rocket:

**Comment by alamb** [comment]
The Apache Arrow Rust community is moving the Rust implementation into its own dedicated github repositories [arrow-rs](https://github.com/apache/arrow-rs) and [arrow-datafusion](https://github.com/apache/arrow-datafusion). It is likely we will not merge this PR into this repository

Please see the [mailing-list](https://lists.apache.org/thread.html/rce7c61cb3f703367dc00984530016e3fcb828e5a8035655fbcccf862%40%3Cdev.arrow.apache.org%3E) thread for more details

We expect the process to take a few days and will follow up with a migration plan for the in-flight PRs.

**Pull request closed without merging**
